src/css3.rs

```diff
@@ -1,120 +1,959 @@
+                    loop {
+                        match self.token_stream.consume() {
+                            Token::Semicolon | Token::Eof => {
+                                // continue
+                            }
+                            _ => {
+                                if let Some(component_value) = self.consume_component_value() {
+                                    tmp.push(component_value);
+                                }
+
+                                // @todo: consume declaration from tmp list
+                            }
+                        }
+                    }
+                }
+                _ => {
+                    // parse error
+                    self.token_stream.reconsume();
+                    self.consume_and_drop_component_values();
+                }
+            }
+        }
+
+        mixed_list
+    }
+
+    fn consume_declaration(&mut self) -> Option<Declaration> {
+        let mut declaration = Declaration::default();
+        let t = self.token_stream.consume();
+        declaration.name = t.to_string();
+
+        self.consume_whitespaces();
+
+        if self.token_stream.lookahead(0) != Token::Colon {
+            // parse error
+            return None;
+        } else {
+            self.token_stream.consume();
+        }
+
+        self.consume_whitespaces();
+
+        loop {
+            match self.token_stream.consume() {
+                Token::Eof => break,
+                _ => {
+                    if let Some(component_value) = self.consume_component_value() {
+                        declaration.value.push(component_value);
+                    }
+                }
+            }
+        }
+
+        if declaration.value.len() >= 2
+            && declaration.value[declaration.value.len() - 2]
+                == ComponentValue::PreservedToken(Token::Delim('!'))
+            && declaration.value[declaration.value.len() - 1]
+                == ComponentValue::PreservedToken(Token::Ident("important".to_string()))
+        {
+            declaration.important = true;
+            declaration.value.pop();
+            declaration.value.pop();
+        }
+
+        while !declaration.value.is_empty()
+            && declaration.value[declaration.value.len() - 1]
+                == ComponentValue::PreservedToken(Token::Whitespace)
+        {
+            declaration.value.pop();
+        }
+
+        Some(declaration)
+    }
+
+    fn consume_component_value(&mut self) -> Option<ComponentValue> {
+        let current_token = self.token_stream.current();
+        match self.token_stream.consume() {
+            Token::LCurly | Token::LBracket | Token::LParen => {
+                match self.consume_simple_block(current_token) {
+                    Some(block) => {
+                        return Some(ComponentValue::SimpleBlock(block));
+                    }
+                    None => {
+                        // parse error
+                    }
```

Comment:
I guess, here it also could be a `if let` especially when you ignore the `None` case. (same for 706-712 and some more, i won't add a comment for all of them)